fix: Allow zero valuation in stock reconciliation

ankush opened the pull request:
Stock reconciliation can not be done for customer provided item as they have zero valuation. This change adds a checkbox in item table to allow such items. The checkbox is automatically checked if item is "Customer provided item".

UI screenshot (checkbox is newly added):
![image](https://user-images.githubusercontent.com/9079960/111167811-7bfb8780-85c7-11eb-8906-4ad63381c4f1.png)


Related issue: ISS-20-21-10248

ankush committed:
fix: Allow zero valuation in stock reconciliation

Stock reconciliation can not be done for customer provided item as they
have zero valuation. This change adds a checkbox in item table to allow
such items.

Related issue: ISS-20-21-10248

sagarvora commented on erpnext/stock/doctype/stock_reconciliation_item/stock_reconciliation_item.json:
```diff
@@ -166,10 +167,19 @@
+   "fieldname": "allow_zero_valuation_rate",
+   "fieldtype": "Check",
+   "label": "Allow Zero Valuation Rate",
+   "no_copy": 1,
```
This may not be needed, since the item itself gets copied.
```suggestion
```

marination commented on erpnext/stock/doctype/stock_reconciliation_item/stock_reconciliation_item.json:
```diff
@@ -166,10 +167,18 @@
    "label": "Batch No",
    "options": "Batch"
+  },
+  {
+   "default": "0",
+   "fieldname": "allow_zero_valuation_rate",
+   "fieldtype": "Check",
+   "label": "Allow Zero Valuation Rate",
+   "print_hide": 1
```
Let's keep this read only and have its display depend on its own value. It is auto set for customer provided items anyway. And as of now for regular items we don't need this.

marination commented on erpnext/stock/doctype/stock_reconciliation/stock_reconciliation.py:
```diff
@@ -31,6 +31,7 @@ def validate(self):
 		self.set_total_qty_and_amount()
 		self.validate_putaway_capacity()
+		self.validate_customer_provided_item()
 
```
Can you also additionally add a function to set a customer provided Item's valuation rate to zero?

In other transactions even if the user accidentally adds a rate, the system automatically sets it to zero again.
Lets do that here to make sure it never accidentally has a valuation.

ankush committed:
fix: set valuation rate for customer items to zero

- In stock reconciliation always set valuation rate of customer provided
items to zero during validation.
- Let user know the valuation has been changed.

ankush committed:
test: customer item in stock reconciliation

marination commented on erpnext/stock/doctype/stock_reconciliation/stock_reconciliation.py:
```diff
@@ -29,6 +29,8 @@ def validate(self):
 		self.validate_data()
 		self.validate_expense_account()
+		self.validate_customer_provided_item()
+		self.set_zero_value_for_customer_items()
```
```suggestion
		self.set_zero_value_for_customer_provided_items()
```

marination commented on erpnext/stock/doctype/stock_reconciliation/stock_reconciliation.py:
```diff
@@ -436,6 +438,19 @@ def validate_expense_account(self):
 			if frappe.db.get_value("Account", self.expense_account, "report_type") == "Profit and Loss":
 				frappe.throw(_("Difference Account must be a Asset/Liability type account, since this Stock Reconciliation is an Opening Entry"), OpeningEntryAccountError)
 
+	def set_zero_value_for_customer_items(self):
```
```suggestion
	def set_zero_value_for_customer_provided_items(self):
```

marination commented on erpnext/stock/doctype/stock_reconciliation/stock_reconciliation.py:
```diff
@@ -436,6 +438,19 @@ def validate_expense_account(self):
+				changed_any_values = True
+
+		if changed_any_values:
+			msgprint(_("Valuation rate for customer provided items has been set to zero."))
```
```suggestion
			msgprint(_("Valuation rate for customer provided items has been set to zero."), title=_("Note"), indicator="blue")
```
